pkg/kubelet/server.go

```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
 		return
 	}
 
-	streamChan := make(chan httpstream.Stream, 1)
-	upgrader := spdy.NewResponseUpgrader()
-	conn := upgrader.UpgradeResponse(response.ResponseWriter, request.Request, func(stream httpstream.Stream) error {
+	podName := kubecontainer.GetPodFullName(pod)
+
+	ServePortForward(response.ResponseWriter, request.Request, s.host, podName, uid, s.host.StreamingConnectionIdleTimeout())
+}
+
+func portForwardStreamReceived(streams chan httpstream.Stream) func(httpstream.Stream) error {
+	return func(stream httpstream.Stream) error {
 		portString := stream.Headers().Get(api.PortHeader)
 		port, err := strconv.ParseUint(portString, 10, 16)
 		if err != nil {
-			return fmt.Errorf("Unable to parse '%s' as a port: %v", portString, err)
+			return fmt.Errorf("unable to parse %q as a port: %v", portString, err)
 		}
 		if port < 1 {
-			return fmt.Errorf("Port '%d' must be greater than 0", port)
+			return fmt.Errorf("port %d must be greater than 0", port)
 		}
-		streamChan <- stream
+		streams <- stream
 		return nil
-	})
+	}
+}
+
+func getPortForwardStreamConnectionID(stream httpstream.Stream) string {
+	connectionID := stream.Headers().Get(api.ConnectionIDHeader)
+	if len(connectionID) == 0 {
+		glog.V(5).Infof("Stream received without connection ID header")
+		// old client that isn't generating connectionID
+		streamID := stream.Identifier()
+
+		// When there aren't concurrent new forwarded connections, each
```

Comment:
if there's an error setting up a stream, it could break the (id/4) rule... I would check the stream type and use Identifier() directly for error streams and Identifier()-2 for data streams